cpp/src/traversal/sssp.cu

```diff
@@ -0,0 +1,430 @@
+  size_t bucket_off_size =
+      ((nnz / TOP_DOWN_EXPAND_DIMX + 1) * NBUCKETS_PER_BLOCK + 2) *
+      sizeof(IndexType);
+  ALLOC_TRY(&exclusive_sum_frontier_vertex_buckets_offsets,
+            bucket_off_size,
+            nullptr);
+
+  // Repurpose d_new_frontier_cnt temporarily
+  IndexType* d_nisolated = d_new_frontier_cnt;
+  cudaMemsetAsync(d_nisolated, 0, sizeof(IndexType), stream);
+
+  // Computing isolated_bmap
+  // Only dependent on graph - not source vertex - done once
+  traversal::flag_isolated_vertices(
+      n, isolated_bmap, row_offsets, vertex_degree, d_nisolated, stream);
+
+  cudaMemcpyAsync(&nisolated,
+                  d_nisolated,
+                  sizeof(IndexType),
+                  cudaMemcpyDeviceToHost,
+                  stream);
+
+  // We need nisolated to be ready to use
+  cudaStreamSynchronize(stream);
+}
+
+template <typename IndexType, typename DistType>
+void SSSP<IndexType, DistType>::configure(DistType* _distances,
+                                          IndexType* _predecessors,
+                                          int* _edge_mask) {
+  distances = _distances;
+  predecessors = _predecessors;
+  edge_mask = _edge_mask;
+
+  useEdgeMask = (edge_mask != NULL);
+  computeDistances = (distances != NULL);
+  computePredecessors = (predecessors != NULL);
+
+  // We need distances for SSSP even if the caller doesn't need them
+  if (!computeDistances)
+    ALLOC_TRY(&distances, n * sizeof(DistType), nullptr);
+}
+
+template <typename IndexType, typename DistType>
+gdf_error SSSP<IndexType, DistType>::traverse(IndexType source_vertex) {
+  // Init distances to infinities
+  traversal::fill_vec(distances, n, traversal::vec_t<DistType>::max, stream);
+
+  // If needed, set all predecessors to non-existent (-1)
+  if (computePredecessors) {
+    cudaMemsetAsync(predecessors, -1, n * sizeof(IndexType), stream);
+  }
+
+  //
+  // Initial frontier
+  //
+
+  cudaMemsetAsync(&distances[source_vertex], 0, sizeof(DistType), stream);
+
+  int current_isolated_bmap_source_vert = 0;
+
+  cudaMemcpyAsync(&current_isolated_bmap_source_vert,
+                  &isolated_bmap[source_vertex / INT_SIZE],
+                  sizeof(int),
+                  cudaMemcpyDeviceToHost);
+
+  // We need current_isolated_bmap_source_vert
+  cudaStreamSynchronize(stream);
+
+  int m = (1 << (source_vertex % INT_SIZE));
+
+  // If source is isolated (zero outdegree), we are done
+  if ((m & current_isolated_bmap_source_vert)) {
+    // Init distances and predecessors are done; stream is synchronized
+    return GDF_SUCCESS;
+  }
+
+  // Adding source_vertex to init frontier
+  cudaMemcpyAsync(&frontier[0],
+                  &source_vertex,
+                  sizeof(IndexType),
+                  cudaMemcpyHostToDevice,
+                  stream);
+
+  // Number of vertices in the frontier and number of out-edges from the
+  // frontier
+  IndexType mf, nf;
+  nf = 1;
+  int iters = 0;
+
+  while (nf > 0) {
+#if 0 
```

Comment:
Better delete temporary code for debugging

Reply:
Ack. Done in next version.